### PR TITLE
add TTCN-3 parser

### PR DESCRIPTION
**Pull request by akrzyz**
TTCN-3 parser based on [patch for ctags 5.8](https://sourceforge.net/p/ctags/patches/55/)

### CODE REVIEW

**Comment by masatake** [comment]
Thank you for contributing universal-ctags.
I check the pull request quickly.
What I found are:

* No copyright notice and license term.
* No test case. Without test cases, we cannot maintain the parser.
  See http://docs.ctags.io/en/latest/units.html#units for writing test cases.
* For windows build, win32/{ctags_vs2013.vcxproj,ctags_vs2013.vcxproj.filters} must be updated.
   What you have to do is simple; just add ttcn.c to some in the files.

**Comment by akrzyz** [comment]
Thank you for comment and guides.
Fixing w32 build shouldn’t be hard.
Writing test cases might be more tough, but I will try to add them somehow in the future.

File: parsers/ttcn.c [hunk 1]
```diff
@@ -0,0 +1,868 @@
...
+    K_ENUM,
+    K_NONE  /* No tag */
+} ttcnKind_t;
+static kindOption ttcnKinds [] = {
+    { TRUE, 'c', "module", "module definition" },
+    { TRUE, 't', "type", "type definition" },
+    { TRUE, 'd', "const", "constant definition" },
+    { TRUE, 'd', "template", "template definition" },
+    { TRUE, 'f', "function", "function definition" },
+    { TRUE, 'f', "signature", "signature definition" },
```
Comment (masatake) [inline parsers/ttcn.c]: Please, use different letter for different things.
How about `s` for signature?

`F` and [0-9] are reserved by ctags.

File: parsers/ttcn.c [hunk 2]
```diff
@@ -0,0 +1,868 @@
...
+    K_PORT,
+    K_MEMBER,
+    K_ENUM,
+    K_NONE  /* No tag */
+} ttcnKind_t;
+static kindOption ttcnKinds [] = {
+    { TRUE, 'c', "module", "module definition" },
+    { TRUE, 't', "type", "type definition" },
+    { TRUE, 'd', "const", "constant definition" },
+    { TRUE, 'd', "template", "template definition" },
```
Comment (masatake) [inline parsers/ttcn.c]: How about `T` ?

File: parsers/ttcn.c [hunk 3]
```diff
@@ -0,0 +1,868 @@
...
+    { TRUE, 'f', "function", "function definition" },
+    { TRUE, 'f', "signature", "signature definition" },
+    { TRUE, 'f', "testcase", "testcase definition" },
```
Comment (masatake) [inline parsers/ttcn.c]: How about `C`?

File: parsers/ttcn.c [hunk 4]
```diff
@@ -0,0 +1,868 @@
...
+    { TRUE, 'f', "altstep", "altstep definition" },
```
Comment (masatake) [inline parsers/ttcn.c]: How about `a`?

File: parsers/ttcn.c [hunk 5]
```diff
@@ -0,0 +1,868 @@
...
+    { TRUE, 'c', "group", "group definition" },
+    { TRUE, 'p', "modulepar", "module parameter definition" },
```
Comment (masatake) [inline parsers/ttcn.c]: p?

File: parsers/ttcn.c [hunk 6]
```diff
@@ -0,0 +1,868 @@
...
+    { TRUE, 'v', "var", "variable instance" },
+    { TRUE, 'v', "timer", "timer instance" },
```
Comment (masatake) [inline parsers/ttcn.c]: t?

File: parsers/ttcn.c [hunk 7]
```diff
@@ -0,0 +1,868 @@
...
+    { TRUE, 'm', "port", "port instance" },
```
Comment (masatake) [inline parsers/ttcn.c]: p?

File: parsers/ttcn.c [hunk 8]
```diff
@@ -0,0 +1,868 @@
...
+
+/* Functions forward declarations */
+static void findTTCNKeyword(ttcnToken_t * pTok);
+static void freeToken(ttcnToken_t * pTok);
+int getNonWhiteSpaceChar (void);
+static ttcnToken_t * getToken (void);
+static void ungetToken (void);
+
+/* Check if token is a TTCN-3 keyword or not */
+static void findTTCNKeyword(ttcnToken_t * pTok) {
```
Comment (masatake) [inline parsers/ttcn.c]: Can you use functions and types declared in main/keyword.h as alternative?

File: parsers/ttcn.c [hunk 9]
```diff
@@ -0,0 +1,868 @@
...
+    K_GROUP,
+    K_MODULEPAR,
+    K_VAR,
+    K_TIMER,
+    K_PORT,
+    K_MEMBER,
+    K_ENUM,
+    K_NONE  /* No tag */
+} ttcnKind_t;
+static kindOption ttcnKinds [] = {
```
Comment (masatake) [inline parsers/ttcn.c]: The kind definitions are the heart of ctags.
So let's preview carefully.
Everything other than them can be fiexed later.

File: parsers/ttcn.c [hunk 10]
```diff
@@ -0,0 +1,868 @@
...
+    { TRUE, 'c', "module", "module definition" },
+    { TRUE, 't', "type", "type definition" },
+    { TRUE, 'd', "const", "constant definition" },
+    { TRUE, 'd', "template", "template definition" },
+    { TRUE, 'f', "function", "function definition" },
+    { TRUE, 'f', "signature", "signature definition" },
+    { TRUE, 'f', "testcase", "testcase definition" },
+    { TRUE, 'f', "altstep", "altstep definition" },
+    { TRUE, 'c', "group", "group definition" },
```
Comment (masatake) [inline parsers/ttcn.c]: How about `g` ?

**Comment by masatake** [comment]
Thank you for updating.
I commented randomly. The most important one is the choice of kind letters.

**Comment by masatake** [comment]
Adding scope information will be helpful to users. (optional)

**Comment by masatake** [comment]
(NOTE: chop with VG, fuzz targets are passed.)

**Comment by masatake** [comment]
I updated the timeout value in master branch. Feel free to rebase your pull reqest on it.

**Comment by akrzyz** [comment]
Hello thank you for comments.
I just rebased my fork.
Each kind has unique letter now and appears in units at least once.

File: parsers/ttcn.c [hunk 11]
```diff
@@ -0,0 +1,869 @@
...
+		 ttcnKeywordCount, sizeof (*ttcnKeywords),
+		 ttcnKeywordsCompare);
+    if (k) {
+	pTok->type = k->id;
+	pTok->kind = k->kind;
+    } else
+	pTok->type = T_ID;
+}
+
+ttcnToken_t * pTtcnToken = NULL;
```
Comment (masatake) [inline parsers/ttcn.c]: static modifier is needed.

File: parsers/ttcn.c [hunk 12]
```diff
@@ -0,0 +1,869 @@
...
+int repeatLastToken = FALSE;
```
Comment (masatake) [inline parsers/ttcn.c]: static modifier is needed.

File: parsers/ttcn.c [hunk 13]
```diff
@@ -0,0 +1,869 @@
...
+
+static void freeToken(ttcnToken_t * pTok) {
+    if(pTok) {
+        if(pTok->value) vStringDelete(pTok->value);
+        eFree (pTok);
+    }
+}
+
+/* This function skips all whitespace and comments */
+int getNonWhiteSpaceChar (void) {
```
Comment (masatake) [inline parsers/ttcn.c]: static modifier is needed.

File: parsers/ttcn.c [hunk 14]
```diff
@@ -0,0 +1,869 @@
...
+    {T_VERDICTTYPE, "verdicttype",  K_NONE},
+    {T_WHILE,       "while",        K_NONE},
+    {T_WITH,        "with",         K_NONE},
+    {T_XOR,         "xor",          K_NONE},
+    {T_XOR4B,       "xor4b",        K_NONE}
+};
+const int ttcnKeywordCount = ARRAY_SIZE(ttcnKeywords);
+
+/* TTCN double-char operators */
+struct s_ttcnOp {
```
Comment (masatake) [inline parsers/ttcn.c]: static modifier is needed.

File: parsers/ttcn.c [hunk 15]
```diff
@@ -0,0 +1,869 @@
...
+    {T_OP_LE,       "<="},
+    {T_OP_GE,       ">="},
+    {T_OP_ASS,      ":="},
+    {T_OP_ARROW,    "->"},
+    {T_OP_SHL,      "<<"},
+    {T_OP_SHR,      ">>"},
+    {T_OP_ROTL,     "<@"},
+    {T_OP_ROTR,     "@>"}
+};
+const int ttcnOpCount = ARRAY_SIZE(ttcnOps);
```
Comment (masatake) [inline parsers/ttcn.c]: static modifier may be needed here.

File: parsers/ttcn.c [hunk 16]
```diff
@@ -0,0 +1,869 @@
...
+    T_UNIVERSAL, T_UNMAP, T_VALUE, T_VALUEOF, T_VAR, T_VARIANT, T_VERDICTTYPE,
+    T_WHILE, T_WITH, T_XOR, T_XOR4B,
+    /* Double-char operators (single-char operators are returned "as is") */
+    T_OP_TO /* .. */, T_OP_EQ /* == */, T_OP_NE /* != */,
+    T_OP_LE /* <= */, T_OP_GE /* >= */,    T_OP_ASS /* := */, T_OP_ARROW /* -> */,
+    T_OP_SHL /* << */, T_OP_SHR /* >> */, T_OP_ROTL /* <@ */, T_OP_ROTR /* @> */
+} ttcnTokenType_t;
+
+/* TTCN keywords. List MUST be sorted in alphabetic order!!! */
+struct s_ttcnKeyword {
```
Comment (masatake) [inline parsers/ttcn.c]: static ?

File: parsers/ttcn.c [hunk 17]
```diff
@@ -0,0 +1,869 @@
...
+    {T_VALUEOF,     "valueof",      K_NONE},
+    {T_VAR,         "var",          K_VAR},
+    {T_VARIANT,     "variant",      K_NONE},
+    {T_VERDICTTYPE, "verdicttype",  K_NONE},
+    {T_WHILE,       "while",        K_NONE},
+    {T_WITH,        "with",         K_NONE},
+    {T_XOR,         "xor",          K_NONE},
+    {T_XOR4B,       "xor4b",        K_NONE}
+};
+const int ttcnKeywordCount = ARRAY_SIZE(ttcnKeywords);
```
Comment (masatake) [inline parsers/ttcn.c]: static

File: parsers/ttcn.c [hunk 18]
```diff
@@ -0,0 +1,869 @@
...
+    vString * value;    /* Semantic value (T_ID and T_LITERAL only) */
+    ttcnKind_t kind;    /* Corresponding tag kind (keywords only) */
+} ttcnToken_t;
+
+/*    LEXER    */
+
+/* Functions forward declarations */
+static void findTTCNKeyword(ttcnToken_t * pTok);
+static void freeToken(ttcnToken_t * pTok);
+int getNonWhiteSpaceChar (void);
```
Comment (masatake) [inline parsers/ttcn.c]: static

File: parsers/ttcn.c [hunk 19]
```diff
@@ -0,0 +1,869 @@
...
+        int brcount = 1;
+        while(brcount > 0) {
+            if(matchToken(br[0])) brcount++;        /* Open */
+            else if(matchToken(br[1])) brcount--;   /* Close */
+            else if(!getToken()) return 0;          /* EOF */
+        }
+        return 1;
+    } else return 0;
+}
+const char BR_CUR[] = "{}";
```
Comment (masatake) [inline parsers/ttcn.c]: static

File: parsers/ttcn.c [hunk 20]
```diff
@@ -0,0 +1,869 @@
...
+const char BR_PAR[] = "()";
```
Comment (masatake) [inline parsers/ttcn.c]: static

File: parsers/ttcn.c [hunk 21]
```diff
@@ -0,0 +1,869 @@
...
+const char BR_SQ[] = "[]";
```
Comment (masatake) [inline parsers/ttcn.c]: static

File: parsers/ttcn.c [hunk 22]
```diff
@@ -0,0 +1,869 @@
...
+
+/* List of TTCN operators.
+   A dot (.) is not a TTCN operator but it is included to simplify
+   the expression parser. Instead of treating X.Y.Z as a single primary
+   the parser treats it as an expression with 3 primaries and two dots. */
+const ttcnTokenType_t ttcnExprOps[] = {
```
Comment (masatake) [inline parsers/ttcn.c]: static

File: parsers/ttcn.c [hunk 23]
```diff
@@ -0,0 +1,869 @@
...
+    T_OR, T_XOR, T_AND, T_NOT, T_OP_EQ, T_OP_NE, '>', '<', T_OP_GE, T_OP_LE,
+    T_OP_SHL, T_OP_SHR, T_OP_ROTL, T_OP_ROTR, T_OR4B, T_XOR4B, T_AND4B, T_NOT4B,
+    '+', '-', '&', '*', '/', T_MOD, T_REM, '.'
+};
+const int ttcnExprOpCount = ARRAY_SIZE(ttcnExprOps);
```
Comment (masatake) [inline parsers/ttcn.c]: static

**Comment by masatake** [comment]
Thank you for updating the changes.
Finally I have some comments about coding style.
I'm summarizing my idea. Please, wait for awhile.

**Comment by masatake** [comment]
https://github.com/universal-ctags/ctags/blob/criteria-document/docs/criteria.rst

This is old note I wrote. This can be used as a check list when reviewing a patch.
I will add "coding style" to it.

**Comment by akrzyz** [comment]
I added more units for ttcn, now coverage for ttcn.c is about 95%.
I also made some "coding style" changes  in ttcn.c based on how other parses look like. Now it should be more coherent.

**Comment by masatake** [comment]
Can I ask you to squash the all commits for ttcn.c into one commit?

**Comment by masatake** [comment]
Thank you very much.

**Comment by masatake** [comment]
Oh, I forgot one thing. We have to update docs/news.rst.

**Comment by akrzyz** [comment]
Thank you for help and guidance through this pull request.

**Pull request merged**